src/main/java/sirius/biz/storage/layer2/jdbc/SQLProcessBlobsLoop.java

```diff
@@ -0,0 +1,145 @@
+import sirius.kernel.di.std.Part;
+import sirius.kernel.di.std.Register;
+import sirius.kernel.health.Exceptions;
+
+import java.sql.SQLException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implements processing actions on {@link SQLDirectory directories} and {@link SQLBlob blobs}.
+ */
+@Register(framework = SQLBlobStorage.FRAMEWORK_JDBC_BLOB_STORAGE)
+public class SQLProcessBlobsLoop extends ProcessBlobsLoop {
+
+    @Part
+    private OMA oma;
+
+    @Override
+    protected AtomicInteger deleteBlobs() {
+        AtomicInteger numBlobs = new AtomicInteger();
+        oma.select(SQLBlob.class).eq(SQLBlob.DELETED, true).limit(256).iterateAll(blob -> {
+            try {
+                if (Strings.isFilled(blob.getPhysicalObjectKey())) {
+                    blob.getStorageSpace().getPhysicalSpace().delete(blob.getPhysicalObjectKey());
+                }
+
+                oma.delete(blob);
+                numBlobs.incrementAndGet();
+            } catch (Exception e) {
+                Exceptions.handle()
+                          .to(StorageUtils.LOG)
+                          .error(e)
+                          .withSystemErrorMessage("Layer 2/SQL: Failed to finally delete the blob %s (%s) in %s: (%s)",
+                                                  blob.getBlobKey(),
+                                                  blob.getFilename(),
+                                                  blob.getSpaceName())
+                          .handle();
+            }
+        });
+
+        return numBlobs;
+    }
+
+    @Override
+    protected AtomicInteger deleteDirectories() {
+        AtomicInteger numDirectories = new AtomicInteger();
+        oma.select(SQLDirectory.class).eq(SQLDirectory.DELETED, true).limit(256).iterateAll(dir -> {
+            try {
+                propagateDelete(dir);
+                oma.delete(dir);
+            } catch (Exception e) {
+                Exceptions.handle()
+                          .to(StorageUtils.LOG)
+                          .error(e)
+                          .withSystemErrorMessage(
+                                  "Layer 2/SQL: Failed to finally delete the directory %s (%s) in %s: (%s)",
+                                  dir.getId(),
+                                  dir.getName(),
+                                  dir.getSpaceName())
+                          .handle();
+            }
+            numDirectories.incrementAndGet();
+        });
+
+        return numDirectories;
+    }
+
+    @Override
+    protected AtomicInteger processCreatedOrRenamedBlobs() {
+        AtomicInteger numBlobs = new AtomicInteger();
+        oma.select(SQLBlob.class).eq(SQLBlob.CREATED_OR_RENAMED, true).limit(256).iterateAll(blob -> {
+            try {
+                createdOrRenamedHandlers.forEach(handler -> handler.execute(blob));
+                numBlobs.incrementAndGet();
+            } catch (Exception e) {
+                Exceptions.handle()
```

Comment:
move the inside .forEach and add the class name of the handler to the message

Reply:
also, this whole "invoke all handlers" could be a single method in the abstract base class as it is independent to the db